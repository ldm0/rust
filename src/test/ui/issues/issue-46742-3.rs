fn add(a: i32, b: i32) -> i32 {
    a + b
}
fn main() {
    // We shouldn't coerce capturing closure to a function
    let cap = 0;
    let _ = (match "+" {
        "+" => add,
        "-" => |a, b| (a - b + cap) as i32,
        _ => unimplemented!(),
    })(5, 5);
    //~^^^ ERROR 9:16: 9:43: `match` arms have incompatible types [E0308]


    // We shouldn't coerce capturing closure to a non-capturing closure
    let _ = (match "+" {
        "+" => |a, b| (a + b) as i32,
        "-" => |a, b| (a - b + cap) as i32,
        _ => unimplemented!(),
    })(5, 5);
    //~^^^ ERROR 18:16: 18:43: `match` arms have incompatible types [E0308]


    // We shouldn't coerce non-capturing closure to a capturing closure
    let _ = (match "+" {
        "+" => |a, b| (a + b + cap) as i32,
        "-" => |a, b| (a - b) as i32,
        _ => unimplemented!(),
    })(5, 5);
    //~^^^ ERROR 27:16: 27:37: `match` arms have incompatible types [E0308]

    // We shouldn't coerce capturing closure to a capturing closure
    let _ = (match "+" {
        "+" => |a, b| (a + b + cap) as i32,
        "-" => |a, b| (a - b + cap) as i32,
        _ => unimplemented!(),
    })(5, 5);
    //~^^^ ERROR 35:16: 35:43: `match` arms have incompatible types [E0308]
}
