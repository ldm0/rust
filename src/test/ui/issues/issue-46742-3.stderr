error[E0308]: `match` arms have incompatible types
  --> $DIR/issue-46742-3.rs:9:16
   |
LL |       let _ = (match "+" {
   |  _____________-
LL | |         "+" => add,
   | |                --- this is found to be of type `fn(i32, i32) -> i32 {add}`
LL | |         "-" => |a, b| (a - b + cap) as i32,
   | |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected fn item, found closure
LL | |         _ => unimplemented!(),
LL | |     })(5, 5);
   | |______- `match` arms have incompatible types
   |
   = note: expected type `fn(i32, i32) -> i32 {add}`
           found closure `[closure@$DIR/issue-46742-3.rs:9:16: 9:43 cap:_]`

error[E0308]: `match` arms have incompatible types
  --> $DIR/issue-46742-3.rs:18:16
   |
LL |       let _ = (match "+" {
   |  _____________-
LL | |         "+" => |a, b| (a + b) as i32,
   | |                --------------------- this is found to be of type `[closure@$DIR/issue-46742-3.rs:17:16: 17:37]`
LL | |         "-" => |a, b| (a - b + cap) as i32,
   | |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected closure, found a different closure
LL | |         _ => unimplemented!(),
LL | |     })(5, 5);
   | |______- `match` arms have incompatible types
   |
   = note: expected type `[closure@$DIR/issue-46742-3.rs:17:16: 17:37]`
           found closure `[closure@$DIR/issue-46742-3.rs:18:16: 18:43 cap:_]`
   = note: no two closures, even if identical, have the same type
   = help: consider boxing your closure and/or using it as a trait object

error[E0308]: `match` arms have incompatible types
  --> $DIR/issue-46742-3.rs:27:16
   |
LL |       let _ = (match "+" {
   |  _____________-
LL | |         "+" => |a, b| (a + b + cap) as i32,
   | |                --------------------------- this is found to be of type `[closure@$DIR/issue-46742-3.rs:26:16: 26:43 cap:_]`
LL | |         "-" => |a, b| (a - b) as i32,
   | |                ^^^^^^^^^^^^^^^^^^^^^ expected closure, found a different closure
LL | |         _ => unimplemented!(),
LL | |     })(5, 5);
   | |______- `match` arms have incompatible types
   |
   = note: expected type `[closure@$DIR/issue-46742-3.rs:26:16: 26:43 cap:_]`
           found closure `[closure@$DIR/issue-46742-3.rs:27:16: 27:37]`
   = note: no two closures, even if identical, have the same type
   = help: consider boxing your closure and/or using it as a trait object

error[E0308]: `match` arms have incompatible types
  --> $DIR/issue-46742-3.rs:35:16
   |
LL |       let _ = (match "+" {
   |  _____________-
LL | |         "+" => |a, b| (a + b + cap) as i32,
   | |                --------------------------- this is found to be of type `[closure@$DIR/issue-46742-3.rs:34:16: 34:43 cap:_]`
LL | |         "-" => |a, b| (a - b + cap) as i32,
   | |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected closure, found a different closure
LL | |         _ => unimplemented!(),
LL | |     })(5, 5);
   | |______- `match` arms have incompatible types
   |
   = note: expected type `[closure@$DIR/issue-46742-3.rs:34:16: 34:43 cap:_]`
           found closure `[closure@$DIR/issue-46742-3.rs:35:16: 35:43 cap:_]`
   = note: no two closures, even if identical, have the same type
   = help: consider boxing your closure and/or using it as a trait object

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0308`.
